### PR TITLE
Implement support for HTTPS Redirects

### PR DESCRIPTION
**Pull request by spencerhance**
Adds support for using HTTPS Redirects with L7 ELB via a Redirects Config in FrontendConfig
Also includes a new status annotation for the additional UrlMap that needs to be synced

e2e tests will be added in a follow up

### CODE REVIEW

**Comment by spencerhance** [comment]
@freehan can you also take a look?

/assign @freehan

File: pkg/utils/namer/frontendnamer.go [hunk 1]
```diff
@@ -88,6 +90,11 @@ func (ln *V1IngressFrontendNamer) UrlMap() string {
 	return ln.namer.UrlMap(ln.lbName)
 }
 
+// UrlMap implements IngressFrontendNamer.
+func (ln *V1IngressFrontendNamer) RedirectUrlMap() string {
+	return ln.namer.RedirectUrlMap(ln.lbName)
```
Comment (freehan) [inline pkg/utils/namer/frontendnamer.go:95]: I strongly recommend NOT to support this feature in V1 Namer.

File: pkg/utils/namer/interfaces.go [hunk 2]
```diff
@@ -25,6 +25,8 @@ type IngressFrontendNamer interface {
 	TargetProxy(protocol NamerProtocol) string
 	// UrlMap returns the name of the URL Map.
 	UrlMap() string
+	// RedirectUrlMap returns the name of the URL Map.
+	RedirectUrlMap() string
```
Comment (freehan) [inline pkg/utils/namer/interfaces.go:29]:
```
// RedirectUrlMap returns the name of the URL Map and if the namer supports naming redirectUrlMap
RedirectUrlMap() (string, bool)
```

File: pkg/utils/namer/frontendnamer.go (hunk 1 above)
Comment (freehan) [reply pkg/utils/namer/frontendnamer.go:95]: Return an error event. Asking user to recreate ingress and it will come with new naming scheme.

Comment (spencerhance) [reply pkg/utils/namer/frontendnamer.go]: Done

File: pkg/utils/namer/interfaces.go (hunk 2 above)
Comment (spencerhance) [reply pkg/utils/namer/interfaces.go]: Done

**Comment by spencerhance** [comment]
@freehan 

>One more thing, there is one corner case: If someone enabled the HTTP redirect and disable later. Will the RedirectURLMap get trimmed? From the code, I do not think so, right?

Good point, I added in code to delete it when disabled

File: pkg/loadbalancers/url_maps.go [hunk 3]
```diff
@@ -80,6 +81,71 @@ func (l *L7) ensureComputeURLMap() error {
...
+
+	name, namerSupported := l.namer.RedirectUrlMap()
+	expectedMap := t.ToRedirectUrlMap(env, l.Versions().UrlMap)
+
+	key, err := l.CreateKey(name)
+	if err != nil {
+		return err
+	}
+
+	currentMap, err := composite.GetUrlMap(l.cloud, key, l.Versions().UrlMap)
```
Comment (freehan) [inline pkg/loadbalancers/url_maps.go:99]: Might worth avoiding this GET call every time an irrelevant ingress was synced. 

One idea is to check if the the ingress status annotation has the redirectUrlMap

File: pkg/loadbalancers/url_maps.go [hunk 4]
```diff
@@ -80,6 +81,71 @@ func (l *L7) ensureComputeURLMap() error {
...
+	if utils.IgnoreHTTPNotFound(err) != nil {
+		return err
+	}
+
+	if expectedMap == nil {
```
Comment (freehan) [inline pkg/loadbalancers/url_maps.go:104]: add a comment here to say "do not expect to have a RedirectUrlMap"

File: pkg/loadbalancers/url_maps.go [hunk 5]
```diff
@@ -80,6 +81,71 @@ func (l *L7) ensureComputeURLMap() error {
...
+		if currentMap == nil {
+			return nil
+		} else {
+			if err := composite.DeleteUrlMap(l.cloud, key, l.Versions().UrlMap); err != nil {
+				return err
+			}
+		}
+	} else {
```
Comment (freehan) [inline pkg/loadbalancers/url_maps.go:112]: probably no need for the else block

File: pkg/loadbalancers/url_maps.go (hunk 3 above)
Comment (spencerhance) [reply pkg/loadbalancers/url_maps.go:100]: Done

File: pkg/loadbalancers/url_maps.go (hunk 4 above)
Comment (spencerhance) [reply pkg/loadbalancers/url_maps.go:106]: Done

File: pkg/loadbalancers/url_maps.go (hunk 5 above)
Comment (spencerhance) [reply pkg/loadbalancers/url_maps.go]: Done

**Comment by coderaven** [comment]
Good to know this has been merged finally! I'm new to this, when is the expected timeline for this to be available in GKE? And is there any documentation that comes along with it?

**Comment by peterldowns** [comment]
Yes, would really love to use this, when will it be available in GKE?

**Comment by artvichi** [comment]
Is this going to be live at some point in this century ?

**Comment by spencerhance** [comment]
Hi folks, please follow the discussion at https://github.com/kubernetes/ingress-gce/issues/1075

We are doing everything we can to launch this ASAP

File: pkg/loadbalancers/target_proxies.go [hunk 6]
```diff
@@ -34,14 +34,29 @@ const (
 
 // checkProxy ensures the correct TargetHttpProxy for a loadbalancer
 func (l *L7) checkProxy() (err error) {
-	isL7ILB := flags.F.EnableL7Ilb && utils.IsGCEL7ILBIngress(l.runtimeInfo.Ingress)
-	tr := translator.NewTranslator(isL7ILB, l.namer)
+	// Get UrlMap Name, could be the url map or the redirect url map
+	// TODO(shance): move to translator
+	var umName string
```
Comment (rhzs) [inline pkg/loadbalancers/target_proxies.go:39]: Why didn't this code written as follow with less branches?

```go
umName := l.um.Name
if flags.F.EnableFrontendConfig {
  // TODO(shance): check for empty name?
    if l.redirectUm != nil && l.runtimeInfo.FrontendConfig.Spec.RedirectToHttps != nil && l.runtimeInfo.FrontendConfig.Spec.RedirectToHttps.Enabled {
      umName = l.redirectUm.Name
    }
}
```

**Pull request merged**